baseplate/lock.py

```diff
@@ -0,0 +1,162 @@
+"""Simple locks in memcached or redis."""
+from __future__ import absolute_import
+
+import os
+import random
+import socket
+import time
+
+from pymemcache.client.base import PooledClient
+from redis import StrictRedis
+
+
+class TimeoutExpired(Exception):
+    pass
+
+
+MIN_RETRY_WAIT = 0.1
+MAX_RETRY_WAIT = 1.0
+
+
+class WrappedMemcache(object):
+    def __init__(self, cache):
+        self.cache = cache
+
+    def add_if_unset(self, key, value, ttl):
+        did_add = self.cache.add(key, value, expire=ttl, noreply=False)
+        return did_add
+
+    def get(self, key):
+        return self.cache.get(key)
+
+    def delete(self, key):
+        self.cache.delete(key, noreply=False)
+
+
+class WrappedRedis(object):
+    def __init__(self, cache):
+        self.cache = cache
+
+    def add_if_unset(self, key, value, ttl):
+        did_add = self.cache.set(key, value, nx=True, ex=ttl)
+
+        # StrictRedis returns True for 'set', and None for 'not set'
+        return bool(did_add)
+
+    def get(self, key):
+        return self.cache.get(key)
+
+    def delete(self, key):
+        self.cache.delete(key)
+
+
+class CacheLock(object):
+    """A lock implemented on memcached or redis.
+
+    Use the cache operation to add a key only if it doesn't exist. If
+    the key is not present in the cache its value is set and the server
+    responds that it was set. If the key is already present in the cache its
+    value is not modified and the server responds that it was not set.
+
+    We use this operation to ensure that only one client can obtain the lock
+    at a time. The first client gets the "set" response and interprets that to
+    mean that the lock was acquired. Any subsequent clients will get the
+    "not set" response and interpret that to mean that some other process has
+    the lock. They can wait and try again, or raise an exception.
+
+    """
+
+    def __init__(self, wrapped_cache, key, hold_limit, wait_limit):
+        self.wrapped_cache = wrapped_cache
+        self.key = key
+        self.hold_limit = hold_limit
+        self.wait_limit = wait_limit
+        self.have_lock = False
+
+    def acquire(self):
+        """Attempt to acquire the lock.
+
+        CacheLocks also support the `context manager protocol`_, for use with
+        Python's ``with`` statement. When the context is entered, the lock
+        calls :py:meth:`acquire` and when the context is exited it
+        automatically calls :py:meth:`release`.
+
+        :raises: :py:exc:`TimeoutExpired` if ``wait_limit`` is exceeded
+            attempting to get the lock.
+
+        .. _context manager protocol:
+            https://docs.python.org/3/reference/datamodel.html#context-managers
+
+        """
+        self.nonce = "%s.%s" % (socket.gethostname(), os.getpid())
```

Comment:
Should we stick a timestamp in here as well?  I think any concurrent requests will all have the same nonce, that might be an issue when deleting if you have a case where your lock expired and another concurrent request in the same app process also acquired the lock and adding a timestamp _should_ address that.